In [1]:
import numpy as np
import pandas as pd

# Load du df depuis BQ

In [2]:
from google.cloud import bigquery
from google.oauth2 import service_account
import sys; sys.path.append('..')
from params import PROJECT_ID, TABLE_ID

def load_dataset(PROJECT_ID, TABLE_ID):

    print(PROJECT_ID, TABLE_ID)

    #TABLE_ID = 'test_dataset.test10'
    TABLE_ID = 'wagon_give_me_a_sign_dataset.hands200'

    credentials = service_account.Credentials.from_service_account_file("../bq_keys.json")
    #credentials = service_account.Credentials.from_service_account_file('~//Users//admin/code//tloizel//wagon-give-me-a-sign//bq_keys.json')

    client = bigquery.Client(project=PROJECT_ID, credentials=credentials)
    table = client.get_table(TABLE_ID)


    # Construct the SQL query to retrieve the table data
    query = f'SELECT * FROM `{table}`'

    # Submit the query and fetch the results
    df = client.query(query).to_dataframe()
    
    return df

In [3]:
df = load_dataset(PROJECT_ID, TABLE_ID)
df.head(3)

wagon-give-me-a-sign wagon_give_me_a_sign_dataset.hands200


x_0  y_0  z_0  x_1  y_1  z_1  x_2  y_2  z_2  x_3  ...  x_18  y_18  z_18  \
0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...   NaN   NaN   NaN   
1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...   NaN   NaN   NaN   
2  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...   NaN   NaN   NaN   

   x_19  y_19  z_19  x_20  y_20  z_20  target  
0   NaN   NaN   NaN   NaN   NaN   NaN       r  
1   NaN   NaN   NaN   NaN   NaN   NaN       r  
2   NaN   NaN   NaN   NaN   NaN   NaN       r  

[3 rows x 64 columns]

In [4]:
df[df['target'] == 'a']

x_0  y_0  z_0       x_1       y_1       z_1       x_2       y_2  \
1319   NaN  NaN  NaN       NaN       NaN       NaN       NaN       NaN   
1320   NaN  NaN  NaN       NaN       NaN       NaN       NaN       NaN   
1321   NaN  NaN  NaN       NaN       NaN       NaN       NaN       NaN   
1322   NaN  NaN  NaN       NaN       NaN       NaN       NaN       NaN   
1323   NaN  NaN  NaN       NaN       NaN       NaN       NaN       NaN   
...    ...  ...  ...       ...       ...       ...       ...       ...   
35179  0.0  0.0  0.0  0.060750  0.003922 -0.010504  0.106186 -0.053520   
35180  0.0  0.0  0.0  0.070393 -0.050977  0.004420  0.109079 -0.127940   
35181  0.0  0.0  0.0  0.054775 -0.010850 -0.012614  0.098121 -0.074492   
35182  0.0  0.0  0.0  0.050755 -0.029991 -0.000021  0.082928 -0.077850   
35183  0.0  0.0  0.0  0.051271 -0.005017 -0.014009  0.094342 -0.062626   

            z_2       x_3  ...      x_18      y_18      z_18      x_19  \
1319        NaN       NaN  ...       NaN       NaN       NaN       NaN   
1320        NaN       NaN  ...       NaN       NaN       NaN       NaN   
1321        NaN       NaN  ...       NaN       NaN       NaN       NaN   
1322        NaN       NaN  ...       NaN       NaN       NaN       NaN   
1323        NaN       NaN  ...       NaN       NaN       NaN       NaN   
...         ...       ...  ...       ...       ...       ...       ...   
35179 -0.006561  0.129712  ...  0.041228 -0.177476  0.055965  0.039056   
35180  0.011604  0.121209  ...  0.014969 -0.269447 -0.015552  0.019862   
35181 -0.013988  0.111118  ...  0.015080 -0.177441 -0.025075  0.010981   
35182  0.003392  0.098258  ...  0.011566 -0.195948 -0.007461  0.012726   
35183 -0.013398  0.111477  ...  0.024271 -0.188487 -0.004318  0.015739   

           y_19      z_19      x_20      y_20      z_20  target  
1319        NaN       NaN       NaN       NaN       NaN       a  
1320        NaN       NaN       NaN       NaN       NaN       a  
1321        NaN       NaN       NaN       NaN       NaN       a  
1322        NaN       NaN       NaN       NaN       NaN       a  
1323        NaN       NaN       NaN       NaN       NaN       a  
...         ...       ...       ...       ...       ...     ...  
35179 -0.155757  0.059135  0.035458 -0.139497  0.061264       a  
35180 -0.219243 -0.010894  0.018529 -0.183170 -0.002473       a  
35181 -0.121619 -0.016015  0.005590 -0.106156 -0.005760       a  
35182 -0.155724 -0.000381  0.010072 -0.127668  0.007668       a  
35183 -0.147909 -0.001305  0.016036 -0.121278  0.006203       a  

[1600 rows x 64 columns]

In [5]:
df[df['target'] == 'b']

x_0  y_0  z_0       x_1       y_1       z_1       x_2       y_2  \
3272   NaN  NaN  NaN       NaN       NaN       NaN       NaN       NaN   
3273   NaN  NaN  NaN       NaN       NaN       NaN       NaN       NaN   
3274   NaN  NaN  NaN       NaN       NaN       NaN       NaN       NaN   
3275   NaN  NaN  NaN       NaN       NaN       NaN       NaN       NaN   
3276   NaN  NaN  NaN       NaN       NaN       NaN       NaN       NaN   
...    ...  ...  ...       ...       ...       ...       ...       ...   
39670  0.0  0.0  0.0  0.053070 -0.023185 -0.026798  0.093336 -0.138699   
39671  0.0  0.0  0.0  0.045187 -0.068694 -0.011558  0.065196 -0.163812   
39672  0.0  0.0  0.0  0.061717 -0.025746 -0.025964  0.103302 -0.149696   
39673  0.0  0.0  0.0  0.058055 -0.051530 -0.016210  0.096042 -0.148981   
39674  0.0  0.0  0.0  0.056619 -0.041428 -0.018423  0.097899 -0.133546   

            z_2       x_3  ...      x_18      y_18      z_18      x_19  \
3272        NaN       NaN  ...       NaN       NaN       NaN       NaN   
3273        NaN       NaN  ...       NaN       NaN       NaN       NaN   
3274        NaN       NaN  ...       NaN       NaN       NaN       NaN   
3275        NaN       NaN  ...       NaN       NaN       NaN       NaN   
3276        NaN       NaN  ...       NaN       NaN       NaN       NaN   
...         ...       ...  ...       ...       ...       ...       ...   
39670 -0.031898  0.070692  ...  0.006516 -0.348074 -0.033773  0.018364   
39671 -0.011755  0.032185  ... -0.019466 -0.300823 -0.007256 -0.019018   
39672 -0.028188  0.089739  ...  0.026217 -0.344985 -0.024871  0.040821   
39673 -0.014644  0.075712  ...  0.031917 -0.339016  0.004300  0.043723   
39674 -0.019629  0.085068  ...  0.035788 -0.302250 -0.014623  0.047418   

           y_19      z_19      x_20      y_20      z_20  target  
3272        NaN       NaN       NaN       NaN       NaN       b  
3273        NaN       NaN       NaN       NaN       NaN       b  
3274        NaN       NaN       NaN       NaN       NaN       b  
3275        NaN       NaN       NaN       NaN       NaN       b  
3276        NaN       NaN       NaN       NaN       NaN       b  
...         ...       ...       ...       ...       ...     ...  
39670 -0.403897 -0.040605  0.027841 -0.455555 -0.045039       b  
39671 -0.344235 -0.013059 -0.020621 -0.387276 -0.020266       b  
39672 -0.395594 -0.032562  0.050613 -0.446418 -0.038983       b  
39673 -0.387159  0.000846  0.054639 -0.434122 -0.003237       b  
39674 -0.347225 -0.021713  0.055119 -0.392875 -0.028602       b  

[1600 rows x 64 columns]

In [35]:
df[df['target'] == 'c']

x_0  y_0  z_0       x_1       y_1       z_1       x_2       y_2  \
2264   NaN  NaN  NaN       NaN       NaN       NaN       NaN       NaN   
2265   NaN  NaN  NaN       NaN       NaN       NaN       NaN       NaN   
2266   NaN  NaN  NaN       NaN       NaN       NaN       NaN       NaN   
2267   NaN  NaN  NaN       NaN       NaN       NaN       NaN       NaN   
2268   NaN  NaN  NaN       NaN       NaN       NaN       NaN       NaN   
...    ...  ...  ...       ...       ...       ...       ...       ...   
37474  0.0  0.0  0.0  0.058182 -0.052335  0.049751  0.096390 -0.115351   
37475  0.0  0.0  0.0  0.058185 -0.050058  0.020275  0.097099 -0.106298   
37476  0.0  0.0  0.0  0.066180 -0.043491  0.039955  0.107669 -0.097539   
37477  0.0  0.0  0.0  0.056934 -0.060653  0.010121  0.094470 -0.144848   
37478  0.0  0.0  0.0  0.055553 -0.063904  0.029336  0.086969 -0.123108   

            z_2       x_3  ...      x_18      y_18      z_18      x_19  \
2264        NaN       NaN  ...       NaN       NaN       NaN       NaN   
2265        NaN       NaN  ...       NaN       NaN       NaN       NaN   
2266        NaN       NaN  ...       NaN       NaN       NaN       NaN   
2267        NaN       NaN  ...       NaN       NaN       NaN       NaN   
2268        NaN       NaN  ...       NaN       NaN       NaN       NaN   
...         ...       ...  ...       ...       ...       ...       ...   
37474  0.062492  0.129396  ...  0.089642 -0.253597 -0.059311  0.125955   
37475  0.025964  0.130719  ...  0.068546 -0.269229 -0.023718  0.100917   
37476  0.053146  0.144636  ...  0.106417 -0.278028 -0.007854  0.143262   
37477  0.014105  0.125618  ...  0.033391 -0.308087 -0.028754  0.063630   
37478  0.040441  0.114805  ...  0.061955 -0.272693 -0.014424  0.091402   

           y_19      z_19      x_20      y_20      z_20  target  
2264        NaN       NaN       NaN       NaN       NaN       c  
2265        NaN       NaN       NaN       NaN       NaN       c  
2266        NaN       NaN       NaN       NaN       NaN       c  
2267        NaN       NaN       NaN       NaN       NaN       c  
2268        NaN       NaN       NaN       NaN       NaN       c  
...         ...       ...       ...       ...       ...     ...  
37474 -0.265422 -0.056364  0.152062 -0.261127 -0.053515       c  
37475 -0.277573 -0.025764  0.128103 -0.267766 -0.026290       c  
37476 -0.291024 -0.001587  0.171416 -0.289317  0.002638       c  
37477 -0.333193 -0.033072  0.092058 -0.331895 -0.033867       c  
37478 -0.285716 -0.013842  0.115265 -0.282494 -0.013278       c  

[1600 rows x 64 columns]

In [36]:
df.columns

Index(['x_0', 'y_0', 'z_0', 'x_1', 'y_1', 'z_1', 'x_2', 'y_2', 'z_2', 'x_3',
       'y_3', 'z_3', 'x_4', 'y_4', 'z_4', 'x_5', 'y_5', 'z_5', 'x_6', 'y_6',
       'z_6', 'x_7', 'y_7', 'z_7', 'x_8', 'y_8', 'z_8', 'x_9', 'y_9', 'z_9',
       'x_10', 'y_10', 'z_10', 'x_11', 'y_11', 'z_11', 'x_12', 'y_12', 'z_12',
       'x_13', 'y_13', 'z_13', 'x_14', 'y_14', 'z_14', 'x_15', 'y_15', 'z_15',
       'x_16', 'y_16', 'z_16', 'x_17', 'y_17', 'z_17', 'x_18', 'y_18', 'z_18',
       'x_19', 'y_19', 'z_19', 'x_20', 'y_20', 'z_20', 'target'],
      dtype='object')

In [37]:
list(df.columns[0:4])

['x_0', 'y_0', 'z_0', 'x_1']

# Autre option : load depuis csv en local

In [11]:
df = pd.read_csv('../csv/coord_live_abcdefghijklmnopqrstuvwxyz.csv', sep = ',' )
df.shape


(3900, 64)

In [12]:
df.head(3)

x_0  y_0  z_0       x_1       y_1       z_1       x_2       y_2       z_2  \
0  0.0  0.0  0.0  0.067701 -0.043059 -0.013783  0.116625 -0.112220 -0.018773   
1  0.0  0.0  0.0  0.061455 -0.026629 -0.019453  0.118439 -0.095989 -0.025251   
2  0.0  0.0  0.0  0.073134 -0.043769 -0.017741  0.129270 -0.115775 -0.023245   

        x_3  ...      x_18      y_18      z_18      x_19      y_19      z_19  \
0  0.139905  ...  0.002343 -0.251392 -0.030343  0.007545 -0.188142 -0.019205   
1  0.141798  ... -0.001150 -0.247533 -0.032880  0.004090 -0.179724 -0.023359   
2  0.154982  ...  0.003378 -0.277963 -0.024795  0.007469 -0.206901 -0.013835   

       x_20      y_20      z_20  target  
0  0.000437 -0.155193 -0.005219       a  
1 -0.001019 -0.142552 -0.009493       a  
2  0.001835 -0.172018  0.001015       a  

[3 rows x 64 columns]

## Preprocessing, formatage de y

In [38]:
from sklearn.model_selection import train_test_split

In [39]:
def balance_df(df):
    if len(df.groupby('target')['target'].count().unique()) > 1:
        print("Not balanced")
        print(df.groupby('target')['target'].count().unique())
        min_ = min(df.groupby('target')['target'].count().unique())
        balanced_df = pd.DataFrame()
        for letter in df['target'].unique():
            sub_df = df[df['target']==letter]
            selected_row = sub_df.sample(n=min_)
            balanced_df = pd.concat([balanced_df,selected_row])
    else:
        print("Balanced")
        print(df.groupby('target')['target'].count().unique())
        balanced_df = df

    print(balanced_df.shape)
    return balanced_df

def shuffle_targets(df):
    shuffled_df = pd.DataFrame()
    for letter in df['target'].unique():
        sub_df = df[df['target']==letter]
        sub_df = sub_df.sample(frac=1)
        shuffled_df = pd.concat([shuffled_df,sub_df])
        #shuffled_df = shuffled_df.reset_index(drop=True)

    return shuffled_df

def train_test_df(df, test_size=0.3, random_state=42):

    df = df.reset_index(drop=True)

    X_train_df = pd.DataFrame()
    X_test_df = pd.DataFrame()
    y_train_df = pd.DataFrame()
    y_test_df = pd.DataFrame()

    for letter in df['target'].unique():
        sub_df = df[df['target']==letter]
        X = sub_df.drop('target', axis=1)
        y = sub_df['target']
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=test_size, random_state=random_state)
        X_train_df = pd.concat([X_train_df, X_train])
        X_test_df = pd.concat([X_test_df, X_test])
        y_train_df = pd.concat([y_train_df, y_train])
        y_test_df = pd.concat([y_test_df, y_test])

    return X_train_df, X_test_df, y_train_df, y_test_df

def preproc(df, test_size=0.3, random_state=42):
    """
    Remove duplicates
    Remove the first three colmns x_0, y_0, z_0
    Balance dataset if needed
    Return X_train_df, X_test_df, y_train_df, y_test_df
    """

    # Remove duplicates
    df = df.dropna()
    df = df.drop_duplicates()

    # Remove the first three columns
    if 'x_0' in df.columns:
        df = df.drop(['x_0','y_0','z_0'], axis=1)

    # check balance and balance it
    balanced_df = balance_df(df)

    # # shuffle par target
    # shuffled_df = shuffle_targets(balanced_df)
    X_train_df, X_test_df, y_train_df, y_test_df = train_test_df(balanced_df, test_size=test_size, random_state=random_state)

    return X_train_df, X_test_df, y_train_df, y_test_df




## On va d'abord encoder la target

In [26]:
y= df['target']
y.head()

0    r
1    r
2    r
3    r
4    r
Name: target, dtype: object

In [27]:
df['target'].unique()

array(['r', 'u', 'i', 'n', 'g', 'z', 't', 's', 'a', 'f', 'o', 'h', 'm',
       'j', 'c', 'd', 'v', 'q', 'x', 'e', 'b', 'k', 'l', 'y', 'p', 'w'],
      dtype=object)

In [28]:
from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()
y_encoded = encoder.fit_transform(y)
y_encoded

array([17, 17, 17, ..., 22, 22, 22])

In [33]:
len(set(y_encoded))

26

In [34]:
df_encoded = df.drop(columns=['target'])
df_encoded.head(3)

x_0  y_0  z_0  x_1  y_1  z_1  x_2  y_2  z_2  x_3  ...  z_17  x_18  y_18  \
0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...   NaN   NaN   NaN   
1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...   NaN   NaN   NaN   
2  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...   NaN   NaN   NaN   

   z_18  x_19  y_19  z_19  x_20  y_20  z_20  
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2   NaN   NaN   NaN   NaN   NaN   NaN   NaN  

[3 rows x 63 columns]

In [45]:
df_encoded['target'] = y_encoded
df_encoded.head(3)

x_0  y_0  z_0  x_1  y_1  z_1  x_2  y_2  z_2  x_3  ...  x_18  y_18  z_18  \
0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...   NaN   NaN   NaN   
1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...   NaN   NaN   NaN   
2  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...   NaN   NaN   NaN   

   x_19  y_19  z_19  x_20  y_20  z_20  target  
0   NaN   NaN   NaN   NaN   NaN   NaN      17  
1   NaN   NaN   NaN   NaN   NaN   NaN      17  
2   NaN   NaN   NaN   NaN   NaN   NaN      17  

[3 rows x 64 columns]

# Train Test Split

In [46]:
X_train_df, X_test_df, y_train_df, y_test_df = preproc(df_encoded, test_size=0.3, random_state=42)

Not balanced
[1437 1449 1371 1453 1414 1406 1369 1370 1425 1394 1428 1418 1419 1395
 1457 1348 1215 1409 1372 1426 1403 1386 1353]
(31590, 61)


In [47]:
X_train_df.shape

(22100, 60)

In [48]:
from tensorflow.lite.keras.models import Sequential
from tensorflow.lite.keras.layers import Dense, LSTM, Dropout
from tensorflow.lite.keras.callbacks import EarlyStopping

# RESULTAT DE CE MODELE : ACCURACY @ 98% et LOSS @ 0.07, 25 minutes pour l'entrainement

# Remarque : Même modèle que pour les lettres A, B, C

# Avec la methode de collecte lag, on a de moins bon resultats : 96% pour loss @ 0.18

In [37]:
#def search_best_model_params(X_train, y_train, X_test, y_test):
    
thierry_model_LSTM_1 = Sequential()

thierry_model_LSTM_1.add(LSTM(128, return_sequences=True, activation='relu', input_shape=(60,1)))
thierry_model_LSTM_1.add(LSTM(64, return_sequences=True, activation='relu'))
thierry_model_LSTM_1.add(LSTM(32, return_sequences=True, activation='relu'))
thierry_model_LSTM_1.add(LSTM(26, return_sequences=False, activation='relu'))
thierry_model_LSTM_1.add(Dropout(0.10))
thierry_model_LSTM_1.add(Dense(26, activation='relu'))
thierry_model_LSTM_1.add(Dropout(0.2))
thierry_model_LSTM_1.add(Dense(16, activation='relu'))
thierry_model_LSTM_1.add(Dropout(0.3))
thierry_model_LSTM_1.add(Dense(26, activation='softmax'))

#model.compile(loss='mse', optimizer=Adam(learning_rate=0.5), metrics=['mae'])
thierry_model_LSTM_1.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

es = EarlyStopping(patience=7, restore_best_weights=True)

#model.compile(loss='mse', optimizer=Adam(learning_rate=0.5), metrics=['mae'])
thierry_model_LSTM_1.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

es = EarlyStopping(patience=10, restore_best_weights=True)

history = thierry_model_LSTM_1.fit(
    X_train_df, y_train_df, 
    #validation_data=(X_test, y_test),
    validation_split = 0.3,
    callbacks=[es],
    epochs=200, batch_size=32, verbose=1)

thierry_model_LSTM_1.evaluate(X_test_df, y_test_df, verbose = 0)
    
    #return params
    

Epoch 1/200
484/484 [==============================] - 100s 188ms/step - loss: 3.0532 - accuracy: 0.0582 - val_loss: 5.1922 - val_accuracy: 0.0000e+00
Epoch 2/200
484/484 [==============================] - 104s 215ms/step - loss: 2.7145 - accuracy: 0.1244 - val_loss: 9.6064 - val_accuracy: 0.0000e+00
Epoch 3/200
484/484 [==============================] - 103s 212ms/step - loss: 2.2189 - accuracy: 0.2143 - val_loss: 14.6068 - val_accuracy: 0.0000e+00
Epoch 4/200
484/484 [==============================] - 100s 207ms/step - loss: 1.8454 - accuracy: 0.3209 - val_loss: 21.6197 - val_accuracy: 0.0000e+00
Epoch 5/200
484/484 [==============================] - 101s 209ms/step - loss: 3940.6824 - accuracy: 0.1961 - val_loss: 3109.4824 - val_accuracy: 1.5081e-04
Epoch 6/200
484/484 [==============================] - 113s 233ms/step - loss: 20098.9883 - accuracy: 0.0559 - val_loss: 4.9730 - val_accuracy: 0.0000e+00
Epoch 7/200
484/484 [==============================] - 98s 203ms/step - loss: 8789

[3.5948235988616943, 0.03846153989434242]

In [23]:
#def search_best_model_params(X_train, y_train, X_test, y_test):
    
thierry_model_LSTM_2 = Sequential()

thierry_model_LSTM_2.add(LSTM(100, return_sequences=True, activation='relu', input_shape=(60,1)))  # première couche LSTM
thierry_model_LSTM_2.add(Dropout(0.2))  # régularisation par dropout
thierry_model_LSTM_2.add(Dense(100, activation='relu'))  # une couche Dense pour le traitement après le LSTM
thierry_model_LSTM_2.add(Dropout(0.2))  # régularisation par dropout
thierry_model_LSTM_2.add(Dense(26, activation='softmax'))  # couche de sortie - 26 pour le nombre de lettres dans l'alphabet

#model.compile(loss='mse', optimizer=Adam(learning_rate=0.5), metrics=['mae'])
thierry_model_LSTM_2.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

es = EarlyStopping(patience=7, restore_best_weights=True)

#model.compile(loss='mse', optimizer=Adam(learning_rate=0.5), metrics=['mae'])
thierry_model_LSTM_2.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

es = EarlyStopping(patience=10, restore_best_weights=True)

history = thierry_model_LSTM_2.fit(
    X_train_df, y_train_df, 
    #validation_data=(X_test, y_test),
    validation_split = 0.3,
    callbacks=[es],
    epochs=200, batch_size=32, verbose=1)

thierry_model_LSTM_2.evaluate(X_test_df, y_test_df, verbose = 0)
    
    #return params

Epoch 1/200


InvalidArgumentError: Graph execution error:

Detected at node 'sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/assert_equal_1/Assert/Assert' defined at (most recent call last):
    File "/Users/admin/.pyenv/versions/3.10.6/lib/python3.10/runpy.py", line 196, in _run_module_as_main
      return _run_code(code, main_globals, None,
    File "/Users/admin/.pyenv/versions/3.10.6/lib/python3.10/runpy.py", line 86, in _run_code
      exec(code, run_globals)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/ipykernel_launcher.py", line 17, in <module>
      app.launch_new_instance()
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/traitlets/config/application.py", line 1043, in launch_instance
      app.start()
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/ipykernel/kernelapp.py", line 725, in start
      self.io_loop.start()
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/tornado/platform/asyncio.py", line 195, in start
      self.asyncio_loop.run_forever()
    File "/Users/admin/.pyenv/versions/3.10.6/lib/python3.10/asyncio/base_events.py", line 600, in run_forever
      self._run_once()
    File "/Users/admin/.pyenv/versions/3.10.6/lib/python3.10/asyncio/base_events.py", line 1896, in _run_once
      handle._run()
    File "/Users/admin/.pyenv/versions/3.10.6/lib/python3.10/asyncio/events.py", line 80, in _run
      self._context.run(self._callback, *self._args)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/ipykernel/kernelbase.py", line 513, in dispatch_queue
      await self.process_one()
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/ipykernel/kernelbase.py", line 502, in process_one
      await dispatch(*args)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/ipykernel/kernelbase.py", line 409, in dispatch_shell
      await result
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/ipykernel/kernelbase.py", line 729, in execute_request
      reply_content = await reply_content
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/ipykernel/ipkernel.py", line 422, in do_execute
      res = shell.run_cell(
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/ipykernel/zmqshell.py", line 540, in run_cell
      return super().run_cell(*args, **kwargs)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3009, in run_cell
      result = self._run_cell(
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3064, in _run_cell
      result = runner(coro)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/IPython/core/async_helpers.py", line 129, in _pseudo_sync_runner
      coro.send(None)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3269, in run_cell_async
      has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3448, in run_ast_nodes
      if await self.run_code(code, result, async_=asy):
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3508, in run_code
      exec(code_obj, self.user_global_ns, self.user_ns)
    File "/var/folders/br/shkkrcw141jc1v89g_7b299r0000gp/T/ipykernel_78528/3132081516.py", line 21, in <module>
      history = thierry_model_LSTM_2.fit(
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/utils/traceback_utils.py", line 65, in error_handler
      return fn(*args, **kwargs)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/engine/training.py", line 1742, in fit
      tmp_logs = self.train_function(iterator)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/engine/training.py", line 1338, in train_function
      return step_function(self, iterator)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/engine/training.py", line 1322, in step_function
      outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/engine/training.py", line 1303, in run_step
      outputs = model.train_step(data)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/engine/training.py", line 1081, in train_step
      loss = self.compute_loss(x, y, y_pred, sample_weight)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/engine/training.py", line 1139, in compute_loss
      return self.compiled_loss(
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/engine/compile_utils.py", line 265, in __call__
      loss_value = loss_obj(y_t, y_p, sample_weight=sw)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/losses.py", line 142, in __call__
      losses = call_fn(y_true, y_pred)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/losses.py", line 268, in call
      return ag_fn(y_true, y_pred, **self._fn_kwargs)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/losses.py", line 2354, in sparse_categorical_crossentropy
      return backend.sparse_categorical_crossentropy(
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/backend.py", line 5762, in sparse_categorical_crossentropy
      res = tf.nn.sparse_softmax_cross_entropy_with_logits(
Node: 'sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/assert_equal_1/Assert/Assert'
assertion failed: [Condition x == y did not hold element-wise:] [x (sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/Shape_1:0) = ] [32 1] [y (sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/strided_slice:0) = ] [32 60]
	 [[{{node sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/assert_equal_1/Assert/Assert}}]] [Op:__inference_train_function_10465]

In [22]:
#def search_best_model_params(X_train, y_train, X_test, y_test):
    
thierry_model_LSTM_3 = Sequential()

thierry_model_LSTM_3.add(LSTM(64, return_sequences=True, activation='relu', input_shape=(60,1)))
#thierry_model_LSTM_3.add(Dropout(0.1))  
thierry_model_LSTM_3.add(LSTM(128, return_sequences=True, activation='relu'))
#thierry_model_LSTM_3.add(Dropout(0.15))  
thierry_model_LSTM_3.add(Dense(128, activation='relu'))
thierry_model_LSTM_3.add(Dropout(0.2))
thierry_model_LSTM_3.add(Dense(16, activation='relu'))
##thierry_model_LSTM_3.add(Dropout(0.3))
thierry_model_LSTM_3.add(Dense(26, activation='softmax'))

es = EarlyStopping(patience=7, restore_best_weights=True)

#model.compile(loss='mse', optimizer=Adam(learning_rate=0.5), metrics=['mae'])
thierry_model_LSTM_3.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

es = EarlyStopping(patience=10, restore_best_weights=True)

history = thierry_model_LSTM_3.fit(
    X_train_df, y_train_df, 
    #validation_data=(X_test, y_test),
    validation_split = 0.3,
    callbacks=[es],
    epochs=200, batch_size=32, verbose=1)

thierry_model_LSTM_3.evaluate(X_test_df, y_test_df, verbose = 0)
    
    #return params

Epoch 1/200


InvalidArgumentError: Graph execution error:

Detected at node 'sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/assert_equal_1/Assert/Assert' defined at (most recent call last):
    File "/Users/admin/.pyenv/versions/3.10.6/lib/python3.10/runpy.py", line 196, in _run_module_as_main
      return _run_code(code, main_globals, None,
    File "/Users/admin/.pyenv/versions/3.10.6/lib/python3.10/runpy.py", line 86, in _run_code
      exec(code, run_globals)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/ipykernel_launcher.py", line 17, in <module>
      app.launch_new_instance()
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/traitlets/config/application.py", line 1043, in launch_instance
      app.start()
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/ipykernel/kernelapp.py", line 725, in start
      self.io_loop.start()
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/tornado/platform/asyncio.py", line 195, in start
      self.asyncio_loop.run_forever()
    File "/Users/admin/.pyenv/versions/3.10.6/lib/python3.10/asyncio/base_events.py", line 600, in run_forever
      self._run_once()
    File "/Users/admin/.pyenv/versions/3.10.6/lib/python3.10/asyncio/base_events.py", line 1896, in _run_once
      handle._run()
    File "/Users/admin/.pyenv/versions/3.10.6/lib/python3.10/asyncio/events.py", line 80, in _run
      self._context.run(self._callback, *self._args)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/ipykernel/kernelbase.py", line 513, in dispatch_queue
      await self.process_one()
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/ipykernel/kernelbase.py", line 502, in process_one
      await dispatch(*args)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/ipykernel/kernelbase.py", line 409, in dispatch_shell
      await result
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/ipykernel/kernelbase.py", line 729, in execute_request
      reply_content = await reply_content
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/ipykernel/ipkernel.py", line 422, in do_execute
      res = shell.run_cell(
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/ipykernel/zmqshell.py", line 540, in run_cell
      return super().run_cell(*args, **kwargs)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3009, in run_cell
      result = self._run_cell(
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3064, in _run_cell
      result = runner(coro)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/IPython/core/async_helpers.py", line 129, in _pseudo_sync_runner
      coro.send(None)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3269, in run_cell_async
      has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3448, in run_ast_nodes
      if await self.run_code(code, result, async_=asy):
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3508, in run_code
      exec(code_obj, self.user_global_ns, self.user_ns)
    File "/var/folders/br/shkkrcw141jc1v89g_7b299r0000gp/T/ipykernel_78528/3085668489.py", line 22, in <module>
      history = thierry_model_LSTM_3.fit(
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/utils/traceback_utils.py", line 65, in error_handler
      return fn(*args, **kwargs)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/engine/training.py", line 1742, in fit
      tmp_logs = self.train_function(iterator)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/engine/training.py", line 1338, in train_function
      return step_function(self, iterator)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/engine/training.py", line 1322, in step_function
      outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/engine/training.py", line 1303, in run_step
      outputs = model.train_step(data)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/engine/training.py", line 1081, in train_step
      loss = self.compute_loss(x, y, y_pred, sample_weight)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/engine/training.py", line 1139, in compute_loss
      return self.compiled_loss(
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/engine/compile_utils.py", line 265, in __call__
      loss_value = loss_obj(y_t, y_p, sample_weight=sw)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/losses.py", line 142, in __call__
      losses = call_fn(y_true, y_pred)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/losses.py", line 268, in call
      return ag_fn(y_true, y_pred, **self._fn_kwargs)
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/losses.py", line 2354, in sparse_categorical_crossentropy
      return backend.sparse_categorical_crossentropy(
    File "/Users/admin/.pyenv/versions/3.10.6/envs/wagon-give-me-a-sign/lib/python3.10/site-packages/keras/src/backend.py", line 5762, in sparse_categorical_crossentropy
      res = tf.nn.sparse_softmax_cross_entropy_with_logits(
Node: 'sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/assert_equal_1/Assert/Assert'
assertion failed: [Condition x == y did not hold element-wise:] [x (sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/Shape_1:0) = ] [32 1] [y (sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/strided_slice:0) = ] [32 60]
	 [[{{node sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/assert_equal_1/Assert/Assert}}]] [Op:__inference_train_function_8099]

# Sauvegarde du modèle

In [23]:
import pickle

def upload_model(model, name_of_model):
    with open(f'{name_of_model}', 'wb') as file:
        pickle.dump(model, file)


def load_model(name_of_the_model=False):
    if name_of_the_model == False:
        with open('model.randomfo', 'rb') as file:
            loaded_model = pickle.load(file)
        return loaded_model
    else:
        with open(name_of_the_model, 'rb') as file:
            loaded_model = pickle.load(file)
        return loaded_model

In [24]:
#upload_model(thierry_model_LSTM, 'thierry_model_LSTM')
upload_model(thierry_model_LSTM_all, 'thierry_model_LSTM_live_all_98_no_verbose')

# Test avec d'autres dataset

In [25]:
#Il faut renseigner le nim du dataset ici

#df_verif = load_dataset(PROJECT_ID, 'test_dataset.test7')
df_verif = load_dataset(PROJECT_ID, 'test_dataset.test10')
#df_verif = pd.read_csv('../csv/coord_live_a_b.csv', sep = ',' )
#df_verif.head(3)

wagon-give-me-a-sign test_dataset.test10


In [26]:
X_verif = df_verif.drop(columns=['x_0', 'y_0', 'z_0', 'target'])
y_verif = df_verif['target']
y_verif_encoded = encoder.transform(y_verif)


In [27]:
#model_verif = load_model('thierry_model_LSTM_live')
model_verif2 = load_model('thierry_model_LSTM_live_all_98')
model_verif3 = load_model('thierry_model_LSTM_live_all_98_no_verbose')

In [33]:
model_verif2.evaluate(X_verif, y_verif_encoded, verbose = 0)

[0.21864818036556244, 0.9333333373069763]

In [32]:
model_verif3.evaluate(X_verif, y_verif_encoded, verbose = 0)

[0.6554323434829712, 0.7507936358451843]

# RESULTAT OK : ACCURACY @ 93% avec le dataset7 et le dataset10